src/WakuRlnV2.sol

```diff
@@ -142,25 +144,61 @@ contract WakuRlnV2 is Initializable, OwnableUpgradeable, UUPSUpgradeable {
     )
         external
         onlyValidIdCommitment(idCommitment)
-        onlyValidUserMessageLimit(userMessageLimit)
+        noDuplicateMembers(idCommitment)
     {
-        _register(idCommitment, userMessageLimit);
+        uint32 index;
+        bool reusedIndex;
+        (index, reusedIndex) = _acquireMembership(_msgSender(), idCommitment, userMessageLimit, true);
+
+        _register(idCommitment, userMessageLimit, index, reusedIndex);
     }
 
-    /// Registers a member
+    /// @notice Allows a user to register as a member
     /// @param idCommitment The idCommitment of the member
     /// @param userMessageLimit The message limit of the member
-    function _register(uint256 idCommitment, uint32 userMessageLimit) internal {
-        if (memberExists(idCommitment)) revert DuplicateIdCommitment();
-        if (commitmentIndex >= SET_SIZE) revert FullTree();
+    /// @param membershipsToErase List of expired idCommitments to erase
+    function register(
+        uint256 idCommitment,
+        uint32 userMessageLimit,
+        uint256[] calldata membershipsToErase
+    )
+        external
+        onlyValidIdCommitment(idCommitment)
+        noDuplicateMembers(idCommitment)
+    {
+        for (uint256 i = 0; i < membershipsToErase.length; i++) {
+            uint256 idCommitmentToErase = membershipsToErase[i];
+            MembershipInfo memory mdetails = members[idCommitmentToErase];
+            if (mdetails.userMessageLimit == 0) revert InvalidIdCommitment(idCommitmentToErase);
+            _eraseMembership(_msgSender(), idCommitmentToErase, mdetails);
+            LazyIMT.update(imtData, 0, mdetails.index);
+        }
+
+        uint32 index;
+        bool reusedIndex;
+        (index, reusedIndex) = _acquireMembership(_msgSender(), idCommitment, userMessageLimit, false);
+
+        _register(idCommitment, userMessageLimit, index, reusedIndex);
+    }
+
+    /// @dev Registers a member
+    /// @param idCommitment The idCommitment of the member
+    /// @param userMessageLimit The message limit of the member
+    /// @param index Indicates the index in the merkle tree
+    /// @param reusedIndex indicates whether we're inserting a new element in the merkle tree or updating a existing
+    /// leaf
+    function _register(uint256 idCommitment, uint32 userMessageLimit, uint32 index, bool reusedIndex) internal {
+        if (nextCommitmentIndex >= SET_SIZE) revert FullTree();
```

Comment:
Can this check be done in the two `register` functions alongside their mofidiers? That would allow us to save gas by reverting earlier if the membership set is full, before even calling `_register`.